bookkeeper-server/src/main/java/org/apache/bookkeeper/discover/ZKRegistrationClient.java

```diff
@@ -212,6 +219,90 @@ public ZooKeeper getZk() {
         return getChildren(bookieReadonlyRegistrationPath, null);
     }
 
+    @Override
+    public CompletableFuture<Versioned<BookieServiceInfo>> getBookieServiceInfo(String bookieId) {
+        String pathAsWritable = bookieRegistrationPath + "/" + bookieId;
+        String pathAsReadonly = bookieReadonlyRegistrationPath + "/" + bookieId;
+
+        CompletableFuture<Versioned<BookieServiceInfo>> promise = new CompletableFuture<>();
+        zk.getData(pathAsWritable, false, (int rc, String path, Object o, byte[] bytes, Stat stat) -> {
+            if (KeeperException.Code.OK.intValue() == rc) {
+                try {
+                    BookieServiceInfo bookieServiceInfo = deserializeBookieService(bytes);
+                    promise.complete(new Versioned<>(bookieServiceInfo, new LongVersion(stat.getCversion())));
+                } catch (IOException ex) {
+                    promise.completeExceptionally(KeeperException.create(KeeperException.Code.get(rc), path));
+                    return;
+                }
+            } else if (KeeperException.Code.NONODE.intValue() == rc) {
+                // not found, looking for a readonly bookie
+                zk.getData(pathAsReadonly, false, (int rc2, String path2, Object o2, byte[] bytes2, Stat stat2) -> {
+                    if (KeeperException.Code.OK.intValue() == rc2) {
+                        try {
+                            BookieServiceInfo bookieServiceInfo = deserializeBookieService(bytes2);
+                            promise.complete(new Versioned<>(bookieServiceInfo, new LongVersion(stat2.getCversion())));
+                        } catch (IOException ex) {
+                            promise.completeExceptionally(KeeperException.create(KeeperException.Code.get(rc2), path2));
+                            return;
+                        }
+                    } else if (KeeperException.Code.NONODE.intValue() == rc2) {
+                        // not found as readonly, the bookie is offline
+                        // return an empty BookieServiceInfoStructure
+                        BookieSocketAddress address = null;
+                        try {
+                            address = new BookieSocketAddress(bookieId);
+                        } catch (UnknownHostException err) {
+                            promise.completeExceptionally(KeeperException.create(KeeperException.Code.get(rc2), path2));
+                            return;
+                        }
+                        BookieServiceInfo.Endpoint endpoint = new BookieServiceInfo.Endpoint();
```

Comment:
This information is wrong.
Why are you adding it?

The bookie does not exist, you can't say that it is exposing this endpoint.
It can result in a false configuration